### PR TITLE
Feature/elf

### PR DESCRIPTION
**Pull request by david942j**
module `ELF`
Support show checksec and load got/plt/symbols
Add dependency https://github.com/david942j/rbelftools

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/23)

### CODE REVIEW

File: lib/pwn.rb [hunk 1]
```diff
@@ -13,6 +13,9 @@
 
 include Pwnlib
 
+# No need to type ELF::ELF every time!
+ELF = ::Pwnlib::ELF::ELF
```
Comment (hanhanW) [inline lib/pwn.rb]: Why not include Pwnlib::ELF ?

File: lib/pwnlib/elf/elf.rb [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        plt_sec = @elf_file.section_by_name('.plt')
+        return if plt_sec.nil? # TODO: log.warn
+        rel_sec = @elf_file.section_by_name('.rel.plt') || @elf_file.section_by_name('.rela.plt')
+        return if rel_sec.nil? # -Wl enabled
+        symtab = @elf_file.section_at(rel_sec.header.sh_link)
+        address = plt_sec.header.sh_addr + 0x10 # magic offset, correct in i386/amd64
+        rel_sec.relocations.each do |rel|
+          symbol = symtab.symbol_at(rel.symbol_index)
+          @plt[symbol.name] = address
+          address += 0x10 # magic gap again
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Define a constant instead of magic number.

File: lib/pwnlib/elf/elf.rb [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      def sections_by_types(types)
+        types.map { |type| @elf_file.sections_by_type(type) }.flatten
+      end
+
+      def base_address
+        return 0 if pie?
+        # Find the min of PT_LOAD's p_vaddr
+        @elf_file.segments_by_type(:load)
+                 .map { |seg| seg.header.p_vaddr }
+                 .select { |addr| addr > 0 }
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Why should it be greater than 0?

File: lib/pwn.rb (hunk 1 above)
Comment (david942j) [inline lib/pwn.rb]: I think you're right

File: lib/pwnlib/elf/elf.rb (hunk 3 above)
Comment (david942j) [inline lib/pwnlib/elf/elf.rb]: There're some segments with `p_vaddr` equals zero
I'll survey why and add a comment for it

File: lib/pwnlib/elf/elf.rb [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+      # Load all exist symbols.
+      def load_symbols
+        @symbols = OpenStruct.new
+        @elf_file.each_sections do |section|
+          next unless section.respond_to?(:symbols)
+          section.each_symbols do |symbol|
+            # Don't care symbols without name.
+            next if symbol.name.empty?
+            next if symbol.header.st_value.zero?
+            @symbols[symbol.name] = symbol.header.st_value
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: What should we do if there are many diffenrent symbols with same name?
Should we add comment here if we don't care?

File: lib/pwnlib/elf/elf.rb [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+      # Load all plt symbols.
+      def load_plt
+        # Unlike pwntools-python, which use unicorn emulating instructions to find plt(s).
+        # Here only use section information, which won't find any plt(s) when compile option '-Wl' is enabled.
+        #
+        # The implementation here same as python-pwntools 3.5, and supports i386 and amd64 only.
+        @plt = OpenStruct.new
+        plt_sec = @elf_file.section_by_name('.plt')
+        return if plt_sec.nil? # TODO: log.warn
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: TODO(david942j)

File: lib/pwnlib/elf/elf.rb [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        rel_sec = @elf_file.section_by_name('.rel.plt') || @elf_file.section_by_name('.rela.plt')
+        return if rel_sec.nil? # -Wl enabled
+        symtab = @elf_file.section_at(rel_sec.header.sh_link)
+        address = plt_sec.header.sh_addr + 0x10 # magic offset, correct in i386/amd64
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Define a constant instead of magic number.

File: lib/pwnlib/elf/elf.rb [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        # Unlike pwntools-python, which use unicorn emulating instructions to find plt(s).
+        # Here only use section information, which won't find any plt(s) when compile option '-Wl' is enabled.
+        #
+        # The implementation here same as python-pwntools 3.5, and supports i386 and amd64 only.
+        @plt = OpenStruct.new
+        plt_sec = @elf_file.section_by_name('.plt')
+        return if plt_sec.nil? # TODO: log.warn
+        rel_sec = @elf_file.section_by_name('.rel.plt') || @elf_file.section_by_name('.rela.plt')
+        return if rel_sec.nil? # -Wl enabled
+        symtab = @elf_file.section_at(rel_sec.header.sh_link)
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Add `return if symtab.nil?` here.

File: lib/pwnlib/elf/elf.rb [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        address = plt_sec.header.sh_addr + 0x10 # magic offset, correct in i386/amd64
+        rel_sec.relocations.each do |rel|
+          symbol = symtab.symbol_at(rel.symbol_index)
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Add `next if symbol.nil?` here?

File: lib/pwnlib/elf/elf.rb [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        dynamic = @elf_file.segment_by_type(:dynamic) || @elf.section_by_name('.dynamic')
+        return nil if dynamic.nil? # No dynamic present, might be static-linked.
+        dynamic.tag_by_type(type)
+      end
+
+      # Load got symbols
+      def load_got
+        @got = OpenStruct.new
+        sections_by_types(%i(rel rela)).each do |rel_sec|
+          symtab = @elf_file.section_at(rel_sec.header.sh_link)
```
Comment (hanhanW) [inline lib/pwnlib/elf/elf.rb]: Add `next if symtab.nil?`

File: test/elf/elf_test.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+require 'test_helper'
+
+require 'pwnlib/elf/elf'
+
+class ELFTest < MiniTest::Test
+  def setup
+    @data = File.join(__dir__, '..', 'data')
+    Rainbow.enabled = false
+    @elf = ::Pwnlib::ELF::ELF.new(File.join(@data, 'victim32'), checksec: false)
```
Comment (hanhanW) [inline test/elf/elf_test.rb]: Should we also test victim64 ?

Comment (hanhanW) [inline test/elf/elf_test.rb]: I think it's really suck, just skip it.

File: lib/pwnlib/elf/elf.rb (hunk 9 above)
Comment (david942j) [inline lib/pwnlib/elf/elf.rb]: `next unless symtab.respond_to?(:symbol_at)` will do the magic

File: lib/pwnlib/elf/elf.rb (hunk 3 above)
Comment (david942j) [inline lib/pwnlib/elf/elf.rb]: sorry I'm wrong

**Comment by hanhanW** [comment]
:lgtm:

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/23#-:-KriphZB0TQy5_nj6s0U:bnfp4nl)*

**Comment by hanhanW** [comment]
Reviewed 5 of 7 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/23)*

**Comment by hanhanW** [comment]
LGTM

**Pull request merged**